File: frontend/src/services/tickets.ts
import axios, { type AxiosResponse } from 'axios'

import api from '@/services/api'
import type { ApiResponse } from '@/types/api'
import type {
  ClaimTicketRequest,
  ClaimTicketResponseData,
  CompleteTicketRequest,
  CompleteTicketResponseData,
  SendTicketMessageRequest,
  SendTicketMessageResponseData,
  TicketDetailResponseData,
  TicketMessagesResponseData,
  TicketsListRequest,
  TicketsListResponseData,
} from '@/types/tickets'
import { isMockApiEnabled } from '@/utils/runtime'

export async function getTickets(
  accessToken: string | null,
  params: TicketsListRequest = {},
): Promise<ApiResponse<TicketsListResponseData>> {
  if (isMockApiEnabled()) {
    const { mockGetTickets } = await import('@/mocks')
    return mockGetTickets(accessToken, params)
  }

  return unwrapApiResponse(
    api.get('/tickets', {
      ...authConfig(accessToken),
      params: toTicketQueryParams(params),
    }),
  )
}

export async function getTicketDetail(
  accessToken: string | null,
  ticketId: string,
): Promise<ApiResponse<TicketDetailResponseData>> {
  if (isMockApiEnabled()) {
    const { mockGetTicketDetail } = await import('@/mocks')
    return mockGetTicketDetail(accessToken, ticketId)
  }

  return unwrapApiResponse(
    api.get(`/tickets/${encodeURIComponent(ticketId)}`, authConfig(accessToken)),
  )
}

export async function claimTicket(
  accessToken: string | null,
  ticketId: string,
  payload: ClaimTicketRequest,
): Promise<ApiResponse<ClaimTicketResponseData>> {
  if (isMockApiEnabled()) {
    const { mockClaimTicket } = await import('@/mocks')
    return mockClaimTicket(accessToken, ticketId, payload)
  }

  return unwrapApiResponse(
    api.post(`/tickets/${encodeURIComponent(ticketId)}/claim`, payload, authConfig(accessToken)),
  )
}

export async function completeTicket(
  accessToken: string | null,
  ticketId: string,
  payload: CompleteTicketRequest,
): Promise<ApiResponse<CompleteTicketResponseData>> {
  if (isMockApiEnabled()) {
    const { mockCompleteTicket } = await import('@/mocks')
    return mockCompleteTicket(accessToken, ticketId, payload)
  }

  return unwrapApiResponse(
    api.post(`/tickets/${encodeURIComponent(ticketId)}/complete`, payload, authConfig(accessToken)),
  )
}

export async function getTicketMessages(
  accessToken: string | null,
  ticketId: string,
): Promise<ApiResponse<TicketMessagesResponseData>> {
  if (isMockApiEnabled()) {
    const { mockGetTicketMessages } = await import('@/mocks')
    return mockGetTicketMessages(accessToken, ticketId)
  }

  return unwrapApiResponse(
    api.get(`/tickets/${encodeURIComponent(ticketId)}/messages`, authConfig(accessToken)),
  )
}

export async function sendTicketMessage(
  accessToken: string | null,
  ticketId: string,
  payload: SendTicketMessageRequest,
): Promise<ApiResponse<SendTicketMessageResponseData>> {
  if (isMockApiEnabled()) {
    const { mockSendTicketMessage } = await import('@/mocks')
    return mockSendTicketMessage(accessToken, ticketId, payload)
  }

  return unwrapApiResponse(
    api.post(`/tickets/${encodeURIComponent(ticketId)}/messages`, payload, authConfig(accessToken)),
  )
}

function authConfig(accessToken: string | null) {
  if (!accessToken) return {}
  return {
    headers: {
      Authorization: `Bearer ${accessToken}`,
    },
  }
}

function toTicketQueryParams(params: TicketsListRequest): Record<string, string | number> {
  const queryParams: Record<string, string | number> = {}

  if (params.scope) queryParams.scope = params.scope
  if (params.status) queryParams.status = params.status
  if (params.priority) queryParams.priority = params.priority
  if (params.keyword) queryParams.keyword = params.keyword
  if (params.page) queryParams.page = params.page
  if (params.page_size) queryParams.page_size = params.page_size

  return queryParams
}

async function unwrapApiResponse<TData>(
  request: Promise<AxiosResponse<ApiResponse<TData>>>,
): Promise<ApiResponse<TData>> {
  try {
    const response = await request
    return response.data
  } catch (error) {
    if (axios.isAxiosError<ApiResponse<TData>>(error) && error.response?.data) {
      return error.response.data
    }
    throw error
  }
}

File: backend/src/api/routes/tickets.py
"""Ticket pool routes."""

from __future__ import annotations

from fastapi import Depends, HTTPException, Query, status
from pycore.api import APIRouter
from sqlalchemy.ext.asyncio import AsyncSession

from src.api.deps import get_current_user
from src.db.models import User
from src.db.session import get_db
from src.models.common import ApiEnvelope
from src.models.tickets import (
    ClaimTicketRequest,
    ClaimTicketResponseData,
    CompleteTicketRequest,
    CompleteTicketResponseData,
    SendTicketMessageRequest,
    SendTicketMessageResponseData,
    TicketDetailResponseData,
    TicketMessagesResponseData,
    TicketsListResponseData,
)
from src.repositories.ticket import TicketRepository
from src.services.ticket import (
    InvalidTicketFilter,
    TicketActionForbiddenError,
    TicketAlreadyClaimedError,
    TicketForbiddenError,
    TicketNotFoundError,
    TicketReplyValidationError,
    TicketService,
)

router = APIRouter(prefix="/api/tickets", tags=["tickets"])


@router.get("", response_model=ApiEnvelope[TicketsListResponseData])
async def list_tickets(
    scope: str = Query("pool"),
    ticket_status: str | None = Query(None, alias="status"),
    priority: str | None = Query(None),
    keyword: str | None = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ApiEnvelope[TicketsListResponseData]:
    service = TicketService(TicketRepository(db))
    try:
        if scope == "pool":
            data = await service.list_pool(
                status=ticket_status,
                priority=priority,
                keyword=keyword,
                page=page,
                page_size=page_size,
            )
        elif scope == "mine":
            data = await service.list_mine(
                user=current_user,
                status=ticket_status,
                priority=priority,
                keyword=keyword,
                page=page,
                page_size=page_size,
            )
        else:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="scope 参数不合法",
            )
    except InvalidTicketFilter as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

    return ApiEnvelope(code=200, message="success", data=data)


@router.get("/{ticket_id}/messages", response_model=ApiEnvelope[TicketMessagesResponseData])
async def get_ticket_messages(
    ticket_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ApiEnvelope[TicketMessagesResponseData]:
    service = TicketService(TicketRepository(db))
    try:
        data = await service.list_messages(ticket_id=ticket_id, user=current_user)
    except TicketNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except TicketForbiddenError as exc:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(exc)) from exc

    return ApiEnvelope(code=200, message="success", data=data)


@router.post("/{ticket_id}/messages", response_model=ApiEnvelope[SendTicketMessageResponseData])
async def send_ticket_message(
    ticket_id: str,
    payload: SendTicketMessageRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ApiEnvelope[SendTicketMessageResponseData]:
    service = TicketService(TicketRepository(db))
    try:
        message = await service.send_message(
            ticket_id=ticket_id,
            user=current_user,
            content=payload.content,
            used_assistant_answer_id=payload.used_assistant_answer_id,
        )
    except TicketNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except TicketActionForbiddenError as exc:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(exc)) from exc
    except TicketReplyValidationError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

    return ApiEnvelope(
        code=200,
        message="success",
        data=SendTicketMessageResponseData(message=message),
    )


@router.post("/{ticket_id}/complete", response_model=ApiEnvelope[CompleteTicketResponseData])
async def complete_ticket(
    ticket_id: str,
    payload: CompleteTicketRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ApiEnvelope[CompleteTicketResponseData]:
    service = TicketService(TicketRepository(db))
    try:
        completed_ticket = await service.complete(
            ticket_id=ticket_id,
            user=current_user,
            employee_id=payload.employee_id,
            summary=payload.summary,
        )
    except TicketNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except TicketActionForbiddenError as exc:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(exc)) from exc
    except TicketReplyValidationError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

    return ApiEnvelope(
        code=200,
        message="success",
        data=CompleteTicketResponseData(ticket=completed_ticket),
    )


@router.get("/{ticket_id}", response_model=ApiEnvelope[TicketDetailResponseData])
async def get_ticket_detail(
    ticket_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ApiEnvelope[TicketDetailResponseData]:
    service = TicketService(TicketRepository(db))
    try:
        data = await service.get_detail(ticket_id=ticket_id, user=current_user)
    except TicketNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except TicketForbiddenError as exc:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(exc)) from exc

    return ApiEnvelope(code=200, message="success", data=data)


@router.post("/{ticket_id}/claim", response_model=ApiEnvelope[ClaimTicketResponseData])
async def claim_ticket(
    ticket_id: str,
    payload: ClaimTicketRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ApiEnvelope[ClaimTicketResponseData]:
    if payload.employee_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="employee_id 与当前账号不一致",
        )

    service = TicketService(TicketRepository(db))
    try:
        claimed_ticket = await service.claim(ticket_id=ticket_id, user=current_user)
    except TicketNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except TicketAlreadyClaimedError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

    return ApiEnvelope(
        code=200,
        message="success",
        data=ClaimTicketResponseData(ticket=claimed_ticket),
    )
